pkg/operator/controller/controller.go

```diff
@@ -0,0 +1,378 @@
+	ManifestFactory        *manifests.Factory
+	KubeConfig             *rest.Config
+	CoreDNSImage           string
+	OpenshiftCLIImage      string
+	OperatorReleaseVersion string
+}
+
+// reconciler handles the actual dns reconciliation logic in response to
+// events.
+type reconciler struct {
+	Config
+
+	// client is the kube Client and it will refresh scheme/mapper fields if needed
+	// to detect some resources like ServiceMonitor which could get registered after
+	// the client creation.
+	// Since this controller is running in single threaded mode,
+	// we do not need to synchronize when changing rest scheme/mapper fields.
+	client kclient.Client
+}
+
+// Reconcile expects request to refer to a cluster dns and will do all the work
+// to ensure the cluster dns is in the desired state.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	errs := []error{}
+	result := reconcile.Result{}
+
+	logrus.Infof("reconciling request: %v", request)
+
+	if request.NamespacedName.Name != DefaultDNSController {
+		// Return a nil error value to avoid re-triggering the event.
+		logrus.Errorf("skipping unexpected cluster dns %s", request.NamespacedName.Name)
+		return result, nil
+	}
+	// Get the current dns state.
+	dns := &operatorv1.DNS{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, dns); err != nil {
+		if errors.IsNotFound(err) {
+			// This means the dns was already deleted/finalized and there are
+			// stale queue entries (or something edge triggering from a related
+			// resource that got deleted async).
+			logrus.Infof("cluster dns not found; reconciliation will be skipped for request: %v", request)
+		} else {
+			errs = append(errs, fmt.Errorf("failed to get cluster dns %s: %v", request, err))
+		}
+		dns = nil
+	}
+
+	if dns != nil {
+		// Ensure we have all the necessary scaffolding on which to place dns instances.
+		if err := r.ensureDNSNamespace(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to ensure dns namespace: %v", err))
+		}
+
+		if dns.DeletionTimestamp != nil {
+			// Handle deletion.
+			if err := r.ensureOpenshiftExternalNameServiceDeleted(); err != nil {
+				errs = append(errs, fmt.Errorf("failed to delete external name for openshift service: %v", err))
+			}
+			if err := r.ensureDNSDeleted(dns); err != nil {
+				errs = append(errs, fmt.Errorf("failed to ensure deletion for cluster dns %s: %v", dns.Name, err))
+			}
+
+			if len(errs) == 0 {
+				// Clean up the finalizer to allow the cluster dns to be deleted.
+				updated := dns.DeepCopy()
+				if slice.ContainsString(dns.Finalizers, DNSControllerFinalizer) {
+					updated.Finalizers = slice.RemoveString(updated.Finalizers, DNSControllerFinalizer)
+					if err := r.client.Update(context.TODO(), updated); err != nil {
+						errs = append(errs, fmt.Errorf("failed to remove finalizer from cluster dns %s: %v", dns.Name, err))
+					}
+				}
+			}
+		} else if err := r.enforceDNSFinalizer(dns); err != nil {
```

Comment:
I think this fixes an unreported defect, similar to [BZ#1682922](https://bugzilla.redhat.com/show_bug.cgi?id=1682922) for cluster-ingress-operator: if someone had created the dns without a finalizer, we would have failed to clean up related resources.